# Keyword (2018/09/16)

### Java

* Object Class

  * 자바 API의 모든 클래스와 사용자가 정의한 모든 클래스의 최상위 클래스이다.
  * 모든 자바 클래스들은 Object클래스로부터 상속받는다.
  * Object 클래스의 모든 메서드와 변수는 다른 모든 클래스에서도 사용 가능하다.<br> 

* Instanceof

  * 객체타입을 확인하는데 사용한다.

  * 속성은 연산자이고 형변환이 가능한 지 해당 여부를 true or false로 가르쳐준다.

  * 주로 부모 객체인지 자식 객체인지 확인하는데 씀

  * 현재 참조하고 있는 클래스를 확인할 수 있는 메소드인 getClass를 함께 알아두면 좋다.

    * 예제 코드

      ~~~java
      class A{}
      class B extends A{}
      class instanceofEx01 {
          
          public static void main(String[] args) {
              A a = new A();
              B b = new B();
              
              System.out.println("a instanceof A : " + (a instanceof A));
              System.out.println("b instanceof A : " + (b instanceof A));
              System.out.println("a instanceof B : " + (a instanceof B));
              System.out.println("b instanceof B : " + (b instanceof B));
          }
      }
      
      <결과>
      a instanceof A : true
      b instanceof A : true
      a instanceof B : false
      b instanceof B : true
      ~~~

      <br> 

* Exception

  * 에러를 무시하고 싶거나 에러가 났을 때 그에 맞는 처리를 하고 싶을 때 사용한다.

  * java는 try , catch , throw등을 이용하여 에러를 처리 할 수 있도록 도와준다.

    * try catch란?

      * try는 예외가 발생할지도 모를 코드 블럭을 정의하는 역할을 함. try블록 다음에는 catch절이 이어짐.
      * catch 는 try블록 내부에서 예외가 발생할 경우 호출되는 문장 블록이다.<br> 

    * throw vs throws

      * throw와 throws둘 다 Exception을 발생시킨다는 것에는 차이가 없다. 하지만 둘 사이의 차이점을 살펴본다면

        * throw
          * throw는 억지로 에러를 발생시키고자 할 때 사용한다(현재 메소드, 혹은 상위메소드로)
        * throws
          * throws는 현재 메소드에서 자신을 호출한 상위 메소드로 Exception을 발생시킴

      * 정리하자면 throws 키워드를 사용하는 메소드를 호출한 상위 메소드에선 이러한 에러 처리에 대한 책임을 맡게 되는 것이고 throw는 프로그래머가 사용자 정의 Exception을 강제로 발생시켜 메소드 내에서 예외처리를 수행하는 것이다. 이 때 throw는 메소드 내에서 예외를 강제로 발생시키거나 현재 메소드에게 Exception 정보를 전달할 수 있다.

        * 예제

          ![error](https://jhkim2017.files.wordpress.com/2017/04/3.png)

        * 실행 순서

          1. main의 try문에서 onAction() 메서드 호출

          2. onAction()의 for문 내에서 try-catch문으로 i==6일 경우 Check6Exception()을 발생(throw 사용)

          3. catch문을 통해 예외처리 및 exception 정보를 현재 메서드로 전달할 수 있다.

          4. throws 키워드를 통해 3. 에서 전달한 exception 정보를 main()으로 전달하여 main() 의catch()문 수행

          5. finally() 수행<br> 

    * try-catch를 쓰는 case와 throw를 쓰는 case

      * try-catch 처럼 예외 처리를 할 수있는데 왜 throw를 사용할까? 일단 try-catch는 직접 그 예외를 처리하는 것이고 throw는 예외를 직접 처리하지 않고 상위 메소드에서 처리한다. 즉, 모듈화를 통해 생각해보자면 일일이 모든 메소드에 try-catch를 쓰기 보단 한 개의 메소드에 발생할 수 있는 모든 에러에 대한 catch절을 만들고 나머지 메소드에선 throw를 통해 예외 처리를 따로 하지않게 하는 것이 효율적이라고 생각한다.
